File: programmers/다트-게임.java
class Solution {

    int[] scoreList = new int[3];
    int listIndex = 0;
    int score = 0;

    public int solution(String dartResult) {
        String newDartResult = dartResult.replace("10", "A");

        for(int i = 0; i < newDartResult.length(); i ++ ){
            char ch = newDartResult.charAt(i);

            if(Character.isDigit(ch) || ch == 'A'){
                if(ch == 'A'){
                    score = 10;
                }else{
                    score = Integer.parseInt(String.valueOf(ch));
                }
            }else{
                switch (ch){
                    case 'S':
                        scoreList[listIndex] = score;
                        listIndex += 1;
                        break;
                    case 'D':
                        scoreList[listIndex] = (score * score);
                        listIndex += 1;
                        break;
                    case 'T':
                        scoreList[listIndex] = (score * score * score);
                        listIndex += 1;
                        break;
                    case '*':
                        if(i == 2){
                            scoreList[0] = (scoreList[0] * 2);
                        }else{
                            scoreList[listIndex - 2] = (scoreList[listIndex - 2] * 2);
                            scoreList[listIndex - 1] = (scoreList[listIndex - 1] * 2);
                        }
                        break;
                    case '#':
                        scoreList[listIndex - 1] = (scoreList[listIndex - 1] * (-1));
                }
            }
        }
        return IntStream.of(scoreList).sum();
    }
}

File: python/programmers/다트-게임.py
# 2018 카카오 기출
def solution(dartResult):
    score_list = [0] * 3
    list_index = 0
    dartResult = dartResult.replace("10", "A")

    for i in range(len(dartResult)):

        if str.isdigit(dartResult[i]) or dartResult[i] == "A":
            # 점수가 10인 경우
            if dartResult[i] == "A":
                score = 10
            else:
                score = int(dartResult[i])
        else:
            # 1배
            if dartResult[i] == "S":
                score_list[list_index] = score
                list_index += 1

            # 2배
            elif dartResult[i] == "D":
                score_list[list_index] = (score * score)
                list_index += 1

            # 3배
            elif dartResult[i] == "T":
                score_list[list_index] = (score * score * score)
                list_index += 1

            # 첫 번째 점수의 *는 첫 번째만 두배
            #나머지는 앞과현재 두배
            elif dartResult[i] == "*": 
                if i == 2:
                    score_list[0] = (score_list[0] * 2)
                else:
                    score_list[list_index - 2] = (score_list[list_index - 2] * 2)
                    score_list[list_index - 1] = (score_list[list_index - 1] * 2)

            # 점수 - 로 변경
            elif dartResult[i] == "#":
                score_list[list_index - 1] = (score_list[list_index - 1] * (-1))

    return sum(score_list)
